### PR TITLE
fix(compilation): key compiled plans on data-derived constants, not just structure

### PR DESCRIPTION
**Pull request by ooples**
## The bug

A compiled training plan is keyed on operation names and output shapes, as the code says:

```csharp
// Shape only — NOT identity. Two steps with the same model
// shape produce the same structure hash.
```

That is sound for parameters, which the plan binds by identity, and for constants that never change. It is **not** sound for a constant the step computes *from its input*: the op structure and every shape are identical on each step, so the plan is reused while the baked-in value goes stale. Nothing detects it.

## How it surfaced

A consumer-side cross-entropy implemented `ignore_index` by dividing by a supervised-example count materialised from the target tensor.

| Execution | Result |
|---|---|
| Eager | loss `2.196550` — exactly `log(9)` for a uniform 9-class head; 105 gradient tensors, none non-finite; max abs gradient `0.60` |
| Compiled plan replay | **NaN on step 1** |

Identical expression, identical inputs. The consumer has since expressed that reduction purely in tensor ops, but the hazard remains for any loss that derives a value from its data — and the failure mode is silent, surfacing far from its cause.

## The fix

The plan key now folds in the **values** of leaves that are neither parameters nor produced by an earlier tape entry. Such a constant then correctly misses the cache and recompiles.

Leaves above a size cap (4096 elements) make the tape **uncompilable** rather than compiled with a baked-in assumption — fingerprinting them every step would cost more than compiling saves, and assuming they are step-invariant is the bug being fixed.

### Deliberate boundaries

- **Parameters are excluded**, by identity. Including them would invalidate the plan every step and make compilation pointless.
- **Non-leaf inputs are skipped** — replay recomputes them.
- **The sources-less overload keeps its cheap structure-only behaviour.** That is what `RecordStep` needs: it only asks whether the step shape repeated, and without the source list it cannot distinguish parameters from constants anyway.
- **Signalled by a `bool`, not a sentinel hash.** Any sentinel is a value a real hash could legitimately take, which would silently disable compilation for that tape forever.
- **Element hashing goes through `EqualityComparer<T>.Default`** to avoid boxing every element on a per-step path.

## Tests

Four added in `DataDerivedConstantHashTests`.

Two cover the new behaviour and were verified to **fail against the previous logic** and pass with this change:
- a data-derived constant changes the key
- an oversized constant makes the tape uncompilable

Two pin the boundaries so the fix cannot over-reach:
- parameter values must **not** affect the key
- the structure-only path must stay structure-only

## Verification

| Suite | Result |
|---|---|
| Compilation tests | 715 passed, 0 failed |
| Autodiff / tape / compilation | 1687 passed, 2 failed |
| New regression tests | 4 passed |

The 2 failures are `GpuCpuAutoDifferentialTests.GpuKernel_Matches_Cpu` on `NativeComplexFFTND` and `NativeComplexIFFTNDReal`. They are **pre-existing and unrelated** — reproduced on a clean checkout with this branch stashed. They require actual GPU hardware (`if (!_gpuReady) return;`) so they do not run on the `ubuntu-latest` CI runners. They are genuine kernel defects, not tolerance noise (`NativeComplexFFTND` shows a 1010% relative error), and are reported separately rather than folded into this change.

## Review round

**`InputsOverflow` was a hole in the original fix.** `TryComputeStructureHash` read only the three inline input slots, but `TapeEntry.GetInputsArray()` returns `InputsOverflow` outright when set — for variadic ops (`TensorAddMany`, `Concat`, `Stack`) that array *is* the input list. A data-derived constant in slot 4+ therefore bypassed the check entirely. Now hashed, with `ConstantInAVariadicOverflowSlot_ChangesTheHash` confirmed to fail against the previous logic.

**Leaf digests are memoised per lookup.** This runs on every `ComputeGradients`, and one constant commonly feeds many entries; each occurrence previously re-walked every element. The digest is still folded in at every occurrence, so position stays part of the hash.

**A pre-existing test-isolation bug surfaced and is fixed.** `TestMinForwardElementsOverride` was a process-wide static gating thread-static state. xunit runs collections in parallel, so while any fixture held the compile gate at a tiny value, unrelated tests on other threads had their small tapes become compile-eligible and intermittently ran a replay path they were never written for. It appeared as `GradientTapeFeatureTests.HardSwish_Backward_IsCorrect` failing only in full-suite runs while passing in isolation. Now thread-scoped, matching the state it gates.

**`WithoutSources_HashIsStructureOnly`** now asserts insensitivity to constant values (two structurally identical tapes, different constants, equal hashes) rather than only that the hash is non-zero.

### Verification after review

| Suite | Result |
|---|---|
| Autodiff + tape + compilation | **1871 passed, 2 failed** |

The 2 are the pre-existing GPU FFT parity cases described above — unrelated, hardware-gated, and separately owned.

### CI workflow

The PR-title lint added here skipped itself for any branch named `release-please--*`, which any contributor could exploit to dodge a required check. Removed outright rather than re-gated: with no `pull-request-title-pattern` configured, release-please's default title (`chore(main): release 1.2.3`) is already a valid conventional commit and passes the lint, so the exemption did nothing. Accepted types now derive from `release-please-config.json` (read from the **base** branch, so a PR cannot widen its own rules) instead of a duplicated list, and the unused `pull-requests: read` scope was dropped.

## Summary by CodeRabbit

- **New Features**
  - Added automated validation for pull request titles using Conventional Commit formatting, with actionable feedback for invalid titles.

- **Bug Fixes**
  - Improved compiled plan validation when calculations include constants derived from input data.
  - Prevented stale plans from being reused when relevant constant values change.
  - Safely refuses compilation for inputs containing constants that are too large to validate reliably.
  - Preserved plan reuse when only parameter values change.

### CODE REVIEW

File: .github/workflows/pr-title-lint.yml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  pull_request:
+    types: [opened, edited, reopened, synchronize]
+
+concurrency:
+  group: pr-title-lint-${{ github.event.pull_request.number }}
+  cancel-in-progress: true
+
+permissions:
+  pull-requests: read
+  contents: read
```
[the remark this answers is not in the document]
Comment (ooples) [reply .github/workflows/pr-title-lint.yml:20]: Reduced, though not to `permissions: {}`. Fixing the type-list duplication below meant reading `release-please-config.json`, so the job now needs a checkout.

Net change: `pull-requests: read` dropped (it was genuinely unused -- the title arrives in the event payload), `contents: read` kept and now documented as being for that single sparse checkout, with `persist-credentials: false`.

Addressed in acbbf3d.

File: .github/workflows/pr-title-lint.yml [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+jobs:
+  lint-title:
+    name: Validate PR title
+    runs-on: ubuntu-latest
+    # release-please maintains its own PRs; don't lint those.
+    if: ${{ !startsWith(github.head_ref, 'release-please--') }}
```
[the remark this answers is not in the document]
Comment (ooples) [reply .github/workflows/pr-title-lint.yml]: Good catch -- removed the exemption entirely rather than re-gating it.

Checking why it was there first: there is no `pull-request-title-pattern` in `release-please-config.json`, so release-please uses its default title, `chore(main): release 1.2.3`. That is itself a valid conventional commit and passes this lint. The exemption was not just bypassable, it was unnecessary.

Verified by running the script against the real config: `chore(main): release 0.120.5` is accepted.

Addressed in acbbf3d.

File: .github/workflows/pr-title-lint.yml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    steps:
+      - name: Check conventional-commit format
+        uses: actions/github-script@3a2844b7e9c422d3c10d287c895573f7108da1b3 # v7
+        with:
+          script: |
+            const title = context.payload.pull_request.title || '';
+            // Types kept in sync with release-please-config.json changelog-sections.
+            const types = ['feat','fix','perf','revert','refactor','build','ci','docs','chore','style','test'];
+            const pattern = new RegExp('^(' + types.join('|') + ')(\\([^)]+\\))?!?: .+');
```
[the remark this answers is not in the document]
Comment (ooples) [reply .github/workflows/pr-title-lint.yml]: Done -- the types now come from `release-please-config.json` rather than a second hand-maintained list.

One hardening detail worth flagging: the config is checked out from the **base branch**, not the PR head. Reading it from the head would let a pull request widen the very rules it is being judged by.

The "triggers a release" line in the failure message also derives from the `hidden` flag now, instead of being a third copy of the same knowledge.

Addressed in acbbf3d.

File: src/AiDotNet.Tensors/Engines/Compilation/AutoTrainingCompiler.cs [hunk 4]
```diff
@@ -223,27 +223,147 @@ private static long IncludeTargetIdentity<T>(long hash, Tensor<T> loss, Tensor<T
...
+                {
+                    return false;
+                }
+                if (entry.Output is not null) produced.Add(entry.Output);
+            }
         }
-        return hash;
+
+        return true;
+    }
```
[the remark this answers is not in the document]
Comment (ooples) [reply src/AiDotNet.Tensors/Engines/Compilation/AutoTrainingCompiler.cs:344]: Fixed. Each distinct leaf's digest is now computed once per lookup and memoised in an identity-keyed dictionary, so a constant feeding many entries is walked once rather than once per occurrence.

The digest is still folded into the running hash at **every** occurrence -- only the element walk is memoised, not the contribution -- so a constant moving between inputs still changes the hash.

Addressed in acbbf3d.

File: src/AiDotNet.Tensors/Engines/Compilation/AutoTrainingCompiler.cs [hunk 5]
```diff
@@ -223,27 +223,147 @@ private static long IncludeTargetIdentity<T>(long hash, Tensor<T> loss, Tensor<T
...
+            if (produced is not null && parameters is not null)
+            {
+                if (!TryHashLeafValue(entry.Input0, produced, parameters, ref hash)
+                    || !TryHashLeafValue(entry.Input1, produced, parameters, ref hash)
+                    || !TryHashLeafValue(entry.Input2, produced, parameters, ref hash))
+                {
+                    return false;
+                }
+                if (entry.Output is not null) produced.Add(entry.Output);
+            }
```
[the remark this answers is not in the document]
Comment (ooples) [reply src/AiDotNet.Tensors/Engines/Compilation/AutoTrainingCompiler.cs]: Correct and genuinely blocking -- thank you. `TapeEntry.GetInputsArray()` returns `InputsOverflow` outright when it is set, so for variadic ops it *is* the input list, and reading only `Input0-2` left a hole straight through the check this method exists to perform.

Now hashed. `Input0` is still hashed as well: `DifferentiableOps` sets both `Input0 = inputs[0]` and `InputsOverflow = inputs`, so that slot is double-counted, which is harmless for a hash and cheaper than special-casing.

Added `ConstantInAVariadicOverflowSlot_ChangesTheHash`, which puts a data-derived constant in slot 5 of a `TensorAddMany`. Confirmed it **fails against the previous logic** and passes with the fix -- otherwise it would not be evidence of anything.

Addressed in acbbf3d.

File: tests/AiDotNet.Tensors.Tests/Engines/Autodiff/DataDerivedConstantHashTests.cs [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+        var loss = _engine.ReduceSum(_engine.TensorMultiply(parameter, constant), null);
+
+        // RecordStep hashes without sources: it only needs to notice that the step SHAPE repeated,
+        // and it cannot tell parameters from constants without the source list. That path must keep
+        // its cheap structure-only behaviour.
+        long structureOnly = AutoTrainingCompiler.ComputeStructureHash(tape.Entries, tape.EntryCount);
+
+        Assert.NotEqual(0L, structureOnly);
+        GC.KeepAlive(loss);
+    }
```
[the remark this answers is not in the document]
Comment (ooples) [reply tests/AiDotNet.Tensors.Tests/Engines/Autodiff/DataDerivedConstantHashTests.cs:245]: Agreed, and adopted -- the non-zero check did not test the property the test name claims. It now builds a second tape with identical structure and a different constant and asserts the two hashes are equal.

Addressed in acbbf3d.

File: src/AiDotNet.Tensors/Engines/Compilation/AutoTrainingCompiler.cs [hunk 7]
```diff
@@ -223,27 +223,200 @@ private static long IncludeTargetIdentity<T>(long hash, Tensor<T> loss, Tensor<T
...
+        // per-step hash of up to MaxHashedConstantElements values is a hot-path regression.
+        var comparer = EqualityComparer<T>.Default;
+        for (int i = 0; i < leaf.Length; i++)
+        {
+            var value = leaf[i];
+            valueHash ^= value is null ? 0 : comparer.GetHashCode(value);
+            valueHash *= unchecked((long)0x100000001b3L);
+        }
+
+        return valueHash;
```
[the remark this answers is not in the document]
Comment (ooples) [reply src/AiDotNet.Tensors/Engines/Compilation/AutoTrainingCompiler.cs:438]: Valid and reachable -- fixed, and I confirmed it is not theoretical: a `[1,2]` and a `[2,1]` constant of equal values both broadcast against a `[2,2]` parameter and both yield a `[2,2]` output, so op name, output shape and value sequence were all identical and the two hashed the same.

`ComputeLeafValueHash` now folds in rank followed by each dimension (rank first, so `[2,2]` and `[2,2,1]` cannot coincide by dimension sequence alone).

Added `ConstantWithSameValuesButDifferentShape_ChangesTheHash`, verified to fail against the previous digest and pass with it.

Addressed in f3279d0.

File: src/AiDotNet.Tensors/Engines/Compilation/AutoTrainingCompiler.cs [hunk 8]
```diff
@@ -266,7 +439,22 @@ internal static long ComputeStructureHash<T>(TapeEntryArena<T> entries, int entr
...
+    /// </remarks>
+    [ThreadStatic]
+    private static long? _testMinForwardElementsOverride;
+
+    /// <inheritdoc cref="_testMinForwardElementsOverride"/>
+    internal static long? TestMinForwardElementsOverride
+    {
+        get => _testMinForwardElementsOverride;
+        set => _testMinForwardElementsOverride = value;
+    }
```
[the remark this answers is not in the document]
Comment (ooples) [reply src/AiDotNet.Tensors/Engines/Compilation/AutoTrainingCompiler.cs:476]: Fixed by clearing it in `ResetState` rather than narrowing the doc -- the override is thread-static now, so leaving it set would let a fixture's gate survive into whatever ran next on that thread, which is the same leak the previous commit set out to close.

Checked the call sites first: `ResetState` is only invoked from the two fixtures' constructors and `Dispose` (the other two mentions are comments), and both set the override *after* their ctor's reset, so clearing it here changes nothing for them.

Addressed in f3279d0.
